homeassistant/components/tplink/__init__.py

```diff
@@ -112,13 +112,13 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the TP-Link component."""
     hass.data.setdefault(DOMAIN, {})
 
-    if discovered_devices := await async_discover_devices(hass):
-        async_trigger_discovery(hass, discovered_devices)
-
     async def _async_discovery(*_: Any) -> None:
         if discovered := await async_discover_devices(hass):
             async_trigger_discovery(hass, discovered)
 
+    hass.async_create_background_task(
+        _async_discovery(), "tplink first discovery", eager_start=True
+    )
     hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, _async_discovery)
```

Comment:
ref: https://github.com/home-assistant/core/pull/111787#discussion_r1506902549

I left discovery running again at started for tplink because discovery is not reliable in tplink and we sometimes need to run it twice to get all the devices to update their ips

Reply:
can it happen now that the discovery is still running while the started event fires?

Reply:
That's a good point, it's safer to take it out. The worst case is they have to wait 15 minutes for the IP to get changed.  I think we can improve discovery in the lib in the future anyways

Reply:
removed in 7e75436efade00a3d05be46407a9a0e3ce43747a, tests adjusted

Reply:
Just realized it probably doesn't matter that much anymore because we have DHCP discovery now